services/placement/staged_placement.go

```diff
@@ -0,0 +1,206 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"sync/atomic"
+
+	placementproto "github.com/m3db/m3cluster/generated/proto/placement"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3x/clock"
+)
+
+var (
+	errNilPlacementSnapshotsProto  = errors.New("nil placement snapshots proto")
+	errNoApplicablePlacement       = errors.New("no applicable placement found")
+	errActiveStagedPlacementClosed = errors.New("active staged placement is closed")
+)
+
+type activeStagedPlacement struct {
+	sync.RWMutex
+
+	placements            []services.ServicePlacement
+	nowFn                 clock.NowFn
+	onPlacementsAddedFn   services.OnPlacementsAddedFn
+	onPlacementsRemovedFn services.OnPlacementsRemovedFn
+
+	expiring int32
+	closed   bool
+	doneFn   services.DoneFn
+}
+
+func newActiveStagedPlacement(
+	placements []services.ServicePlacement,
+	opts services.ActiveStagedPlacementOptions,
+) services.ActiveStagedPlacement {
+	p := &activeStagedPlacement{
+		placements:            placements,
+		nowFn:                 opts.ClockOptions().NowFn(),
+		onPlacementsAddedFn:   opts.OnPlacementsAddedFn(),
+		onPlacementsRemovedFn: opts.OnPlacementsRemovedFn(),
+	}
+	p.doneFn = p.onPlacementDone
+
+	if p.onPlacementsAddedFn != nil {
+		p.onPlacementsAddedFn(placements)
+	}
+
+	return p
+}
+
+func (p *activeStagedPlacement) ActivePlacement() (services.ServicePlacement, services.DoneFn, error) {
+	p.RLock()
+	placement, err := p.activePlacementWithLock(p.nowFn().UnixNano())
+	if err != nil {
+		p.RUnlock()
+		return nil, nil, err
+	}
+	return placement, p.doneFn, nil
+}
+
+func (p *activeStagedPlacement) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.closed {
+		return errActiveStagedPlacementClosed
+	}
+	if p.onPlacementsRemovedFn != nil {
+		p.onPlacementsRemovedFn(p.placements)
+	}
+	p.placements = nil
+	return nil
+}
+
+func (p *activeStagedPlacement) onPlacementDone() { p.RUnlock() }
+
+func (p *activeStagedPlacement) activePlacementWithLock(timeNanos int64) (services.ServicePlacement, error) {
+	if p.closed {
+		return nil, errActiveStagedPlacementClosed
+	}
+	idx := p.activeIndexWithLock(timeNanos)
+	if idx < 0 {
+		return nil, errNoApplicablePlacement
+	}
+	placement := p.placements[idx]
+	// If the placement that's in effect is not the first placment, expire the stale ones.
+	if idx > 0 && atomic.CompareAndSwapInt32(&p.expiring, 0, 1) {
+		go p.expire()
+	}
+	return placement, nil
+}
+
+// activeIndexWithLock finds the index of the last placement whose cutover time is no
+// later than t (a.k.a. the active placement). The cutover times of the placements are
+// sorted in ascending order (i.e., earliest time first).
+func (p *activeStagedPlacement) activeIndexWithLock(timeNanos int64) int {
+	idx := 0
+	for idx < len(p.placements) && p.placements[idx].CutoverNanos() <= timeNanos {
```

Comment:
iterate from 0 is going to be faster because in most cases the active placement is the first one, and if it isn't, the expire goroutine will make it so.